mautic/mautic
# SMS enhancements - support callback with received, read and failed notifications

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )

#### Notice for testers and developers

This PR continue work on PR https://github.com/mautic/mautic/pull/7449 it should be merged to staging before test this PR

#### Description:
https://github.com/mautic/mautic/pull/7449 bring support for text messages transport callback and receive info about reply.
This PR extend it with callback about deliveries info and stats (delivered, read, failed) or campaign decision (is delivered, is read, is failed).
This callback support process multiple results from webhook. 

This can be used by any integration transport. 

Every transport can define:
- what callback data can receive (reply, delivered info...)
- what kind of statistic would display (delivered, read, failed) based on TransportSettingsInterface
- add properties to SMS - for example our transport can add device (phone number) to each sms and media file from provider servers (whatsapp)

Stats in sms list 

![image](https://user-images.githubusercontent.com/462477/63173512-518ac680-c040-11e9-9dea-fd20061f7e7f.png)

Stats in detail

![image](https://user-images.githubusercontent.com/462477/63173620-8ac33680-c040-11e9-8334-ca80dac8d648.png)

Custom SMS properties extend by plugin

![image](https://user-images.githubusercontent.com/462477/63223002-646fd900-c1af-11e9-94fe-ea92e46ec6fb.png)


#### Steps to test this PR:
1. You can test Twilio. Should works like before. You can use Mautibox for test http://mautibox.com/7795
1. Contact me, I'll send you custom Text message transports to test another features of PR
